langchain/callbacks/index.ts

```diff
@@ -0,0 +1,267 @@
+  handleLLMError?: (err: string, verbose?: boolean) => void;
+  handleLLMEnd?: (output: LLMResult, verbose?: boolean) => void;
+
+  handleChainStart?: (
+    chain: { name: string },
+    inputs: ChainValues,
+    verbose?: boolean
+  ) => void;
+  handleChainError?: (err: Error, verbose?: boolean) => void;
+  handleChainEnd?: (output: ChainValues, verbose?: boolean) => void;
+
+  handleToolStart?: (
+    tool: { name: string },
+    action: AgentAction,
+    verbose?: boolean
+  ) => void;
+  handleToolError?: (err: Error, verbose?: boolean) => void;
+  handleToolEnd?: (output: string, verbose?: boolean) => void;
+  handleAgentEnd?: (action: AgentFinish, verbose?: boolean) => void;
+}
+
+export const createFromHandlers =
+  (handlers: PerEventCallbackHandler): CallbackHandler =>
+  (event, args) => {
+    switch (event.event) {
+      case "llm.start":
+        return handlers.handleLLMStart?.(
+          event.llm,
+          event.prompts,
+          args?.verbose
+        );
+      case "llm.new_token":
+        return handlers.handleLLMNewToken?.(event.token, args?.verbose);
+      case "llm.error":
+        return handlers.handleLLMError?.(event.err, args?.verbose);
+      case "llm.end":
+        return handlers.handleLLMEnd?.(event.output, args?.verbose);
+      case "chain.start":
+        return handlers.handleChainStart?.(
+          event.chain,
+          event.inputs,
+          args?.verbose
+        );
+      case "chain.error":
+        return handlers.handleChainError?.(event.err, args?.verbose);
+      case "chain.end":
+        return handlers.handleChainEnd?.(event.output, args?.verbose);
+      case "agent.tool_start":
+        return handlers.handleToolStart?.(
+          event.tool,
+          event.action,
+          args?.verbose
+        );
+      case "agent.tool_end":
+        return handlers.handleToolEnd?.(event.output, args?.verbose);
+      case "agent.tool_error":
+        return handlers.handleToolError?.(event.err, args?.verbose);
+      case "agent.end":
+        return handlers.handleAgentEnd?.(event.action, args?.verbose);
+      default:
+        return undefined;
+    }
+  };
+
+export abstract class BaseCallbackHandler extends CallableInstance<
+  Parameters<CallbackHandler>,
+  ReturnType<CallbackHandler>
+> {
+  constructor() {
+    super("handle");
+  }
+
+  abstract handle(
+    event: CallbackEvent,
+    args?: CallbackKwargs
+  ): Promise<void> | void;
+}
+
+export interface BaseCallbackManager extends BaseCallbackHandler {
+  addHandler: (handler: CallbackHandler) => void;
+  removeHandler: (handler: CallbackHandler) => void;
+  setHandlers: (handlers: CallbackHandler[]) => void;
+}
+
+const isEventType = <T extends string>(
+  prefix: T,
+  event: CallbackEvent
+): event is CallbackEvent & { event: `${T}.${string}` } =>
+  event.event.startsWith(prefix);
+
+export abstract class EventTypeCallbackHandler extends BaseCallbackHandler {
```

Comment:
I think we should make all handlers async (and by extension the manager too)

Reply:
I think in most cases they are async already? (returning `Promise<void> | void`)

Looks like i missed it on the `PerEventCallbackHandler` object but should be easy fix